### DISCUSSION TITLE
Strange battery sensors

### DISCUSSION THREAD

**Discussion by bengtner** [wlcrs/huawei_solar#149] [General]
I have had my battery monitored and controlled by this integration for a few weeks now. It works well, but it takes some time to fully understand the behaviour of this device. Still there are a few things that I don't understand. I have configured my energy dashboard to include the battery and used Battery total charge and Battery total discharge as the data source. The output seems to be pretthy accurate for charging. If I charge the battery one hour, approximately 2,5 kwh goes into the battery. However, the last few days I haven't discharged the battery at all. This clearly seen when I examine the Battery Charge Discharge power sensor. I get a positve value during one hour when charging takes place. The rest of the day this sensor shows 0 kw. But when I look on the Energy Dashboard it shows a the battery has discharged 0.8 kwh!  See enclosed pictures.
![image](https://user-images.githubusercontent.com/18145417/195187798-e2980635-1d7b-4db7-9b10-78702d05fff1.png)
![image](https://user-images.githubusercontent.com/18145417/195188726-74f589e6-1773-42c6-96b9-3856fb8bd734.png)

This puzzles me. Am I using the wrong sensors as input for the dashboard? How  come the dashboard show 0.8kwh discharge during the day when the battery charge discharge power sensor only indicate postive values during the day? If I look on data from a day where I actually discharge the battery, the charge and discharge sensor shows both positiv and negative values.

**Comment by bengtner** [comment]
Here comes a screenshot showing a set of entities on one single screen illustrating my problem.
![image](https://user-images.githubusercontent.com/18145417/195191846-1f92e0aa-d5b8-482a-a592-789ee6ab8b22.png)
Edit: I should have selected the original battery mode selector instead of my own variant in the screen dump. Idle means Time of use mode with charge/discharge power set to 0. Discharge means Fully fed to grid. The curve looks fine in this mode, but in Time of use mode it looks like the battery "leaks" energy without showing negative power. This is really strange.

**Comment by bengtner** [comment]
I selected a few more/other  battery parameters and made a new screendump. Here I think you can see what happens. The battery seems to "leak" some energy even when I have set charge/discharge power to 0 during the long period in TOU (Time of use) mode. You notice the battery bus current is -0.1 A a significant part of this time. The explains why I end up with some battery discharge during a day even if my intention was to keep the battery idling. However, this just seems to happen in TOU mode. During "Fully fed to grid" the battery seems to behave differently. Or maybe this is dependant on charging level. 

![image](https://user-images.githubusercontent.com/18145417/195252358-168a1ed1-9460-49e8-bf3c-03372dd8700b.png)

**Comment by wlcrs** [comment]
I've observed the same behavior in my installation. My theory is that keeping the battery and/or inverter online costs a significant amount of energy.

You can also see this when looking to the efficiency charts of the inverter: when producing less than 5% of it's maximum power, the efficiency falls into a cliff, probably due to the non-negligible constant cost in terms of power consumption.

Sadly, I have not found a solution to put the battery into a proper standby mode to prevent this discharging from happening.

**Reply by clunit** [reply]
@wlcrs Have you tried the shutdown command? I've been discussing this with @MichaelPihlblad and he found and tried with the "Inverter ON/OFF" switch in the SUN2000 app, and that seems to work fine to reduce "idle" power consumption. 

It may not be perfect for everyone, we can assume live "Backup box" may not work for example. But I use LUNA and SUN2000 only for energy storage/charge/discharge based on price/consumption hourly so it would be perfect. I've tried the same On/Off switch from FusionSolar Web as well (can't find it in my FusionSolar App?). It puts the inverter in "Standby" mode, and the battery in "Offline" mode (no lights on at all). You can still communicate with the inverter/dongle, and the Smart Meter data is still available. It takes about 2 minutes to get up and running when switching on again.

@MichaelPihlblad found what looks like the Modbus parameters for these commands. It seems they are not implemented in the Python library from what we could see though so it must also be added there. 

![IMG_9643](https://user-images.githubusercontent.com/100047725/208988340-2351e545-9c48-45be-a4b8-e72a151316b5.jpeg)
![IMG_9640](https://user-images.githubusercontent.com/100047725/208988391-dc9aa625-6890-42c5-900b-961187a1b8e1.jpeg)
![IMG_9641](https://user-images.githubusercontent.com/100047725/208988405-1f9eca60-9aa0-4fbd-b960-89e708730458.jpeg)
![IMG_9642](https://user-images.githubusercontent.com/100047725/208988434-4c64d28c-8bbf-4a80-ae9c-19738586d56a.jpeg)

**Reply by wlcrs** [reply]
Thank you for writing this up! These registers were indeed overlooked.

Besides sending the actual startup and shutdown command (for which I'll need to investigate which data needs to be sent to those registers), I also need to be able to detect what the current state of the inverter is to correctly set the status of the toggle switch. Preliminary research suggests this can be detected via register 32089 'Device status'.

Can you please let me know what the reported status of your inverter is when you have switched of the inverter via that switch. You can find this by going in HA to Settings -> Devices -> Huawei Solar -> Inverter -> Diagnostics -> 'Device status'. Is it `Shutdown: manual restart required` by any chance?

![image](https://user-images.githubusercontent.com/2150060/209002430-9f66aa0b-80e5-4bf6-997b-4fe78827d471.png)

**Reply by clunit** [reply]
Great, that sounds promising!

Here's Device status:
<img width="997" alt="image" src="https://user-images.githubusercontent.com/100047725/209007494-ca694952-9271-49c9-a84d-7aa449d89889.png">
So it is "Shutdown: command" during all power off. Then when I power on it's a start sequence which I suppose is the same no matter how it's started: "Standby: no irradiation" > "Standby: initializing" > "Standby: detecting irradiation" > "Standby: detecting insulation resistance". 

Here's the "Inverter state" if you should need it:
<img width="996" alt="image" src="https://user-images.githubusercontent.com/100047725/209007295-97175314-4a3f-417f-b969-ffae6999dde5.png">

**Reply by clunit** [reply]
We also get a "Shutdown time" that might be useful for longer periods of shutdowns maybe:
<img width="997" alt="image" src="https://user-images.githubusercontent.com/100047725/209008451-6e7e9a6f-41b8-4d3c-89d5-208b0db12609.png">

Also the Locking status goes to "Locked", don't know what that is?

**Reply by wlcrs** [reply]
Initial support has been added with https://github.com/wlcrs/huawei_solar/commit/9c4bc9ea1f9284b8f9b8e4f4b8339144cfcc904e .  Note that it's part of a larger update that I've been working on in the past weeks to add "active capacity control". I hope to release it during the holidays.

**Reply by clunit** [reply]
Sounds great! I'm happy to test it out when you feel comfortable. Is it possible to test from the dev-branch or should we wait for the release? (btw, I think the link above is to the wrong commit, but I see the changes in the dev-branch and also the python lib)

**Reply by wlcrs** [reply]
Just checkout the develop branch. I updated it this morning.

**Reply by clunit** [reply]
Tried it quickly last night, but got an error but did not have time to check it further. I noticed the battery had went from "Offline" to "Standby" and just had to get it working for the night. Not much time today either, but tried again tonight and got the following when trying to go from On to Off or the opposite (switched through FusionSolar). The switch seems to detect the correct state, but not able to make the call to make the switch. Not had time to investigate any further at all. Let me know any further info I can provide.

<img width="617" alt="image" src="https://user-images.githubusercontent.com/100047725/209412148-13e9d022-7517-4954-9c12-d98d73f4ead9.png">

Logger: homeassistant.components.websocket_api.http.connection
Source: custom_components/huawei_solar/switch.py:267 
Integration: Home Assistant WebSocket API (documentation, issues) 
First occurred: December 23, 2022 at 5:55:58 PM (3 occurrences) 
Last logged: December 23, 2022 at 11:54:30 PM

[139768194267120] __enter__
[139768301747408] __enter__
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/websocket_api/commands.py", line 200, in handle_call_service
    await hass.services.async_call(
  File "/usr/src/homeassistant/homeassistant/core.py", line 1745, in async_call
    task.result()
  File "/usr/src/homeassistant/homeassistant/core.py", line 1782, in _execute_service
    await cast(Callable[[ServiceCall], Awaitable[None]], handler.job.target)(
  File "/usr/src/homeassistant/homeassistant/helpers/entity_component.py", line 213, in handle_service
    await service.entity_service_call(
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 678, in entity_service_call
    future.result()  # pop exception if have
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 943, in async_request_call
    await coro
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 715, in _handle_entity_call
    await result
  File "/config/custom_components/huawei_solar/switch.py", line 267, in async_turn_off
    with self._change_lock:
AttributeError: __enter__

**Reply by clunit** [reply]
I don't think the battery mode switch has anything to do with Huawei_solar. But on the other hand there is these registers (37000 and 37741) which might be interesting to monitor/log/handle the battery operating mode:
<img width="827" alt="image" src="https://user-images.githubusercontent.com/100047725/209412589-4542be9e-60f1-41e1-8e58-f697e2ef71ff.png">

Tonight the mode has not been switched, it's still "offline" after ~6 hours, but the last two times it switched to "standby" after 30-45 minutes. It's probably some Huawei thing :)  I'm suspecting maybe the operating mode before Off command could be the issue, that's the only thing I know may have been different. Not much time too investigate thoroughly this week..

**Reply by wlcrs** [reply]
> "/config/custom_components/huawei_solar/switch.py", line 267, in async_turn_off with self._change_lock: AttributeError: **enter**

Thank you for the stacktrace, a last minute change before the release of 1.2.0 beta 2 contained a small error. It's been corrected in 1.2.0 beta 3.

**Reply by wlcrs** [reply]
> I don't think the battery mode switch has anything to do with Huawei_solar. But on the other hand there is these registers (37000 and 37741) which might be interesting to monitor/log/handle the battery operating mode: <img alt="image" width="827" src="https://user-images.githubusercontent.com/100047725/209412589-4542be9e-60f1-41e1-8e58-f697e2ef71ff.png">
> 
> Tonight the mode has not been switched, it's still "offline" after ~6 hours, but the last two times it switched to "standby" after 30-45 minutes. It's probably some Huawei thing :) I'm suspecting maybe the operating mode before Off command could be the issue, that's the only thing I know may have been different. Not much time too investigate thoroughly this week..

When creating this integration, I decided to only use the registers that refer to both batteries at once (eg. prefixed by `[Energy storage]`, not by `[Energy storage unit X]`). This was to simplify the implementation and improve performance (easier to query those registers as I can bundle them in 1 request).
It would be very hard to add any of the `[Energy storage unit X]` registers now and stay backward compatible.

**Reply by clunit** [reply]
Ah, Ok! That's probably fine, I just missed that one was available.

I've been running the on/off switch a few times back and forth last few days and it seems to work fine! Only issue seems to be the Huawei-related one (where the battery sometimes goes from Offline to Standby and thereby consumes a bit of energy).

**Reply by wlcrs** [reply]
Just FYI: switching the inverter on and off will physically trigger some relays and can cause short voltage peaks in some components while everything gets started up/shut down and leveled out. This will cause wear on these components, and can lead to premature failure of the inverter. I would therefore advice against excessive use of that feature.

(I'm not an expert in this matter, but my engineering background tells me that it's best to mention this)

**Reply by MichaelPihlblad** [reply]
Should be fine as long as no current is flowing over the relays while opening.

You have a point, one time I rebooted the inverter while charging, I will never do that again, it did not sound good. 

Considering how crappy the Huawei sw is in general it is probably wise to make sure no charging or discharging is ongoing when requesting  the shutdown command. To avoid getting problems with sticky relays etc.

**Comment by bengtner** [comment]
Good to know there are other poeple that have observed this strange behaviour.  It is weird the battery behaves like this. Spending 2.5kwh per day just to keep the battery online seems to be very high cost.

**Comment by MichaelPihlblad** [comment]
I have the exact same problem. I have an ongoing support request to Huawei.

My standby power consumption is 60-80W.  Which is equivalent to > 0.5% of the SOC per hour. > 12% / day.  Really extreme.  Meaning battery fully drained within one week. 

In the datasheet for the inverter they define standby consumption as night mode with < 5.5W consumption. The battery datasheet does not specify any standby power consumption for what I can find. 

I have seen on Huawei forum others having similar issues, adviced to contact support by E-mail.

**Reply by bengtner** [reply]
Ok, I hope you will share any answer from Huawei support here!

**Comment by MichaelPihlblad** [comment]
Ok, Huawei got back. Saying to update firmware. 

They attached to the e-mail firmware updates 146 for inverter and 116 for battery. 

I had 144 in inverter and 114 in battery. Updated just a few weeks ago 

Strange thing is on Huawei portal you can now download even newer SW. 148 for inverter and 119 for battery,  one wonders why support did not send those newer in the E-mail.

Will test now and get back to you with the result.

**Reply by Sheilaarklow** [reply]
Could this be the reason my system seems to drain between 8 to 12 kw twice daily in one go.
That amounts to almost half of the battery storeage of 20kw, 
If the manual says it should be 5kw something is wrong.
the storage area can get hot these days as we have outside temperatures of up to 30degrees.
Should I place a cooling fan in the room?

**Comment by tarvotarvo** [comment]
I also noticed that my battery drains at about 100-110 W rate. About 0.1-0.11 kWh of capacity lost every hour. 
I have 3 modules (15 kWh in total). 
Also the battery control module temperature does not drop below 38 degrees C although the room temperature is 20 degrees C (battery status is 'running' but discharge power set to 0 W).

Also hoping to hear about any improvements after FW upgrades.

**Comment by MichaelPihlblad** [comment]
Sorry to say no improvement after upgrading firmware. Still approx 83W standby consumption.

Support replied they will investigate.

**Reply by tarvotarvo** [reply]
Just curious - do you have 2 battery modules? I am wondering why I have over 100 W and you have slightly less drainage.

**Reply by MichaelPihlblad** [reply]
No I have 3 modules. 15 kwh.

**Comment by MichaelPihlblad** [comment]
Huawei basically seems to state that 150watt standby power consumption is normal. 10 Watt when disconnecting the inverter.  If I managed to interpret correctly. Weird reasoning IMHO


> SUN2000MA
> 
> Normal phenomenon:energy storage and inverter self-consumption.
> 
> 1．Innormal cases, the loss of the inverter in standby mode is about 10 W. However,the power loss of the inverter in the grid-tied state is about 100 W. Inaddition to the energy storage loss, the total power loss is about 150 W.
> 
> 2．The inverter and battery keepin standby mode and do not make up for rest. It's for quick response to theload It's for quick response to the load. It takes about 2 minutes to start theinverter and battery. If it sleep, nighttime energy storage cannot respondquickly to the load. (Power components (switching loss and shutdown loss) existin the inverter and battery. Therefore, loss occurs when the inverter is instandby mode. The energy source of the inverter is either the grid or theenergy storage. Getting energy from the grid costs money, so getting energyfrom batteries.
> 
> [Root Cause] NA
> [Solution]NA
>

**Reply by wlcrs** [reply]
Thank you for sharing this response from Huawei. This is the formal acknowledgement from what I already suspected.

**Comment by stefanhapper** [comment]
What is the best way to identify the exact loss for my own installation (5 kw battery with 4.6 KTL-L1 inverter)?

If I understand the response from Huawei correctly, during Summer when the battery and inverter are always on, we could be constantly losing 150 W (or 3.6 kWh of my 5 kWh battery over 24 hours)? That would be quite astonishing.

**Reply by MichaelPihlblad** [reply]
You can look at "battery state of capacity" and "battery day discharge" for the standby hours to identify the standby consumption.

**Comment by ruteclrp** [comment]
I have noticed that the batteries have built in heaters to keep them at some reasonable internal temperature during cold period. I have also noted that the battery drain is substantially higher during very cold conditions.
The internal battery heating feature is mentioned somewhere in the LUNA documentation (I remember).

**Comment by moffermann223** [comment]
Currently I am figuring out the same. I was wondering why my energy consumption over night from the batteries shown in the Home Assistant Energy Dashboard exceeds the energy consumed over the day with no change of consumers is the house. Over night the battery is discharged with 0,3 - 0,33 kWh every hour (sensor.battery_total_discharge). As soon as the battery gets charging power from the PV or is fully loaded (no discharge anymore), my self-consumed energy drops by approx. 0,1 kWh every hour. This matches pretty much the experiences described above of approx. 100 W drain power.

Where does this energy goes to? Is it the idle-energy and conversion losses consumed by the inverter which I cannot measure/obtain during PV-Production? I can't imagine that the battery is "leaking". Since my inverter and batteries are located in a fairly warm room, battery heating can't be the issue.

**Comment by Sheilaarklow** [comment]
Some answers I have had is the possibility of a malfunctioning or badly
installed  something like a non rreturn valve for water, a device that does
not allow energy to flow back to the pv.  Also incorrectly wired batteries,
and possibly the parametres set on your batteries as to when they should
discharge..
I think there has been such a sureg in demand for solar power that
companies cannot employ enough correctly qualified engineers as it is all
about sales. But soon I think this will be their nemises as problems will
occur and they have not the know how to fix the issues.

**Comment by moffermann223** [comment]
I quickly built a sensor to check the power discrepancy between the dc supply power and the so called "active" power.
I calculated the DC-String power simply with Voltage * Current to know the power comming down from my roof. Voltage and Current is given by the Modbus Sensors. With having that, the following calculation should give information about the "lost" energy needed by the inverter itself:

```
- template: 
  - sensor:
    - name: "PV Inverter Idle Power"
      unit_of_measurement: "W"
      device_class: power
      state_class: measurement 
      icon: mdi:lightning-bolt
      state: >-
          {{(
             (states.sensor.inverter_active_power.state | float) -
             (states.sensor.pv_string_1_power.state | float) -       #self-defined sensor
             (states.sensor.pv_string_2_power.state | float) -      #self-defined sensor
             (states.sensor.battery_charge_discharge_power.state | float) | round (2) 
          )}}
```

In this moment, with  75 % of PV-Peak Power delivery, it shows a value of around 300 W.

**Comment by Sheilaarklow** [comment]
I have no tec background and your abilities are amazing! Thing is we as
customers should not be left having to do this
Where do you live Michael?

**Comment by moffermann223** [comment]
Hey Sheila (?? I am just guessing :-) ),

not that hard as an electrical engineer with some practical hands-on.
I am from Germany.

**Comment by Sheilaarklow** [comment]
Well I am just researching online. Even with that sometimes do feel I know
more than the installers.
Why cannot they take responsibility for their work.

**Reply by moffermann223** [reply]
Well, they do, but they are "just" installers and not the developers of the components itself. And I am even pretty sure that not more than 5 % of their customers understand the background or even trying to create the statistics of what we are doing here.

**Comment by Sheilaarklow** [comment]
Well ,Michael. You said it. Customers. Not investors.
A customer has not the possible privilege of earning from all new
tecnologies.
Customers pay for a product or service and deserve to get what they pay for.
They should not be used as a platform for research and development
Time for Tapas here.
Chat again.
